user-management/server/server.ts

```diff
@@ -0,0 +1,234 @@
+}
+
+const app = express();
+app.use(express.json());
+
+app.post('/api/auth/sign-up', async (req, res, next) => {
+  try {
+    const { username, password } = req.body;
+    if (!username || !password) {
+      throw new ClientError(400, 'username and password are required fields');
+    }
+
+    const hashedPassword = await argon2.hash(password);
+    const sql = `insert into "users" ("username", "hashedPassword")
+                  values($1, $2 )
+                  returning "createdAt","userId","username";`;
+    const params = [username, hashedPassword];
+    const result = await db.query(sql, params);
+    const user = result.rows[0];
+    res.status(201).json(user);
+    /* TODO:
+     * Delete the "Not implemented" error.
+     * Hash the user's password with `argon2.hash()` (note that this method is async)
+     * Insert the user's "username" and "hashedPassword" into the "users" table.
+     * Respond to the client with a 201 status code and the new user's "userId", "username", and "createdAt" timestamp.
+     * Catch any errors.
+     *
+     * Hint: Insert statements can include a `returning` clause to retrieve the inserted row(s).
+     */
+  } catch (err) {
+    next(err);
+  }
+});
+
+// we dont use authMiddleware in sign in or sign up
+app.post('/api/auth/sign-in', async (req, res, next) => {
+  try {
+    // after sending and receiving request body
+    const { username, password } = req.body as Partial<Auth>;
+    // const response = req.get('Authorization'); // key 'content-type' from the request or extracting headers
+    // console.log(response); // check the server terminal for output which is application/json
+    if (!username || !password) {
+      // is this if the body is missing ?
+      throw new ClientError(401, 'missing body request');
+    }
+    // the select returns the userId and hashed password not the username
+    // the username from the body request
+    // the sql is the response to request from client
+    const sql = `select "userId", "hashedPassword" from "users"
+                where "username" = $1;`;
+
+    // the code below after getting the response or sending the response
+    // else if user found
+    // password from body, but the existing password in database
+    const params = [username];
+    const result = await db.query(sql, params);
+    const user = result.rows[0];
+    // now user can access all properties in row[0] in the table rows
+    // like user.userId, user.username, or user.createdAt
+
+    if (!user) {
+      throw new ClientError(401, 'user does not exist');
+    }
+    // else if user exists
+    // or just (argon2.verify(password, result.rows[0])
+    // password is from the body request plain text
+    // argon2.verify returns a promise, so we need to await the result
+    // verifiedPass receives boolean result
+    // or
+    // verify(password, result.rows[0].hashedPassword) , but according to AI it does matter the order
```

Comment:
The order does matter. The verify function signature tells you that the first argument is the hashed password and the second is the clear password.